utils.groovy

```diff
@@ -282,4 +282,110 @@ def get_artifacts_to_build(pipecfg, stream, basearch) {
+                                 credentialsId: "${cloud}-build-upload-config")]) {
+
+            def bucket = pipecfg.clouds."${cloud}".bucket
+            def path = pipecfg.clouds."${cloud}".path
+
+            if (cloud in ['aws', 'aliyun']) {
+                def primary_region = pipecfg.clouds."${cloud}".primary_region
+                def grant_user_args
+                if (cloud == 'aws') {
+                    def aws_accounts = pipecfg.clouds."${cloud}".upload_accounts
+                    grant_user_args = aws_accounts.collect{"--grant-user ${it}"}.join(" ")
+                    bucket = "s3://${bucket}/$path"
+                }
+                shwrap("""
+                cosa buildextend-${cloud} \
+                    --upload \
+                    --arch=${basearch} \
+                    --build=${buildID} \
+                    --region=${primary_region} ${grant_user_args} \
+                    --bucket ${bucket} \
+                    --credentials-file=\${${var_name}_BUILD_UPLOAD_CONFIG}
+                """)
+           }
+           if (cloud == 'azure') {
+               def container = pipecfg.clouds."${cloud}".container
+               def resource_group = pipecfg.clouds."${cloud}".resource_group
+               def storage_account = pipecfg.clouds."${cloud}".storage_account
+               shwrap("""cosa buildextend-azure \
+                   --upload \
+                   --auth \${AZURE_IMAGE_UPLOAD_CONFIG} \
+                   --build=${buildID} \
+                   --container ${container} \
+                   --profile \${AZURE_IMAGE_UPLOAD_CONFIG} \
+                   --resource-group ${resource_group} \
+                   --storage-account ${storage_account} \
+                   --force
+                """)
+           }
+           if (cloud == 'gcp') {
+                def create_image =  (pipecfg.clouds."${cloud}".create_image ? "--create-image=true": "")
+                def deprecated =  (pipecfg.clouds."${cloud}".deprecated ? "--deprecated": "")
+                def description =  (pipecfg.clouds."${cloud}".description ? ('--description=' + pipecfg.clouds."${cloud}".description): "")
+                def family =  (pipecfg.clouds."${cloud}".family ? '--family ' + pipecfg.clouds."${cloud}".family: "")
+                def acl = (pipecfg.clouds."${cloud}".public ? "--public" : "")
+                def gcp_licenses = pipecfg.clouds."${cloud}".licenses
+                def licenses = gcp_licenses.collect{"--license " + it.replace('STREAM', stream)}.join(" ")
+                def today = shwrapCapture("date +%Y-%m-%d")
+                family = family.replace("STREAM", stream)
+                description = description.replace("BUILDID", buildID)
+                description = description.replace("STREAM", stream)
+                description = description.replace("BASEARCH", basearch)
+                description = description.replace("TODAY", today)
```

Comment:
I'm fine with this as is.

But I'd still say that this kind of nontrivial logic ultimately should live in coreos-assembler and not the pipeline.

The GCP licenses in particular to me seems like a strong match for https://github.com/coreos/coreos-assembler/pull/3094

Reply:
+1 I think the idea is to moved it as a followup in the future ;)